src/pid/CtreController.java

```diff
@@ -0,0 +1,110 @@
+package pid;
+
+import static pid.PIDConstants.TIMEOUT;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.IMotorControllerEnhanced;
+import com.ctre.phoenix.motorcontrol.SupplyCurrentLimitConfiguration;
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
+import com.ctre.phoenix.motorcontrol.can.TalonSRX;
+import sensors.counter.CtreEncoder;
+
+public abstract class CtreController extends AbstractController {
+  protected IMotorControllerEnhanced ctreMotorController;
+  protected CtreEncoder ctreEncoder;
+  protected int pidSlot;
+  protected int timeoutMs;
+
+  public CtreController(IMotorControllerEnhanced motorControllerEnhanced, CtreEncoder ctreEncoder) {
+    super(0, 0, 0, 0);
+    this.ctreMotorController = motorControllerEnhanced;
+    this.ctreEncoder = ctreEncoder;
+    this.pidSlot = 0;
+    this.timeoutMs = TIMEOUT;
+  }
+
+  public CtreController(IMotorControllerEnhanced motorControllerEnhanced, CtreEncoder ctreEncoder,
+                        int pidSlot, int timeoutMs) {
+    super(0, 0, 0, 0);
+    this.ctreMotorController = motorControllerEnhanced;
+    this.ctreEncoder = ctreEncoder;
+    this.pidSlot = pidSlot;
+    this.timeoutMs = timeoutMs;
+  }
+
+  public CtreController(IMotorControllerEnhanced motorControllerEnhanced, CtreEncoder ctreEncoder,
+                        double kP, double kI, double kD, double kF, int pidSlot, int timeoutMs) {
+    super(kP, kI, kD, kF);
+    this.ctreMotorController = motorControllerEnhanced;
+    this.ctreEncoder = ctreEncoder;
+    this.pidSlot = pidSlot;
+    this.timeoutMs = timeoutMs;
+  }
+
+  public IMotorControllerEnhanced getCtreMotorController() {
+    return ctreMotorController;
+  }
+
+  public CtreEncoder getCtreEncoder() {
+    return ctreEncoder;
+  }
+
+  public int getTimeoutMs() {
+    return timeoutMs;
+  }
+
+  public void setTimeoutMs(int timeoutMs) {
+    this.timeoutMs = timeoutMs;
+  }
+
+  public int getPidSlot() {
+    return this.pidSlot;
+  }
+
+  public void setPidSlot(int pidSlot) {
+    this.pidSlot = pidSlot;
+  }
+
+  @Override
+  public void setPIDFTerms(double kP, double kI, double kD, double kF) {
+    super.setPIDFTerms(kP, kI, kD, kF);
+    ctreMotorController.config_kP(pidSlot, kP, this.timeoutMs);
+    ctreMotorController.config_kI(pidSlot, kI, this.timeoutMs);
+    ctreMotorController.config_kD(pidSlot, kD, this.timeoutMs);
+    ctreMotorController.config_kF(pidSlot, kF, this.timeoutMs);
+  }
+
+  public void setRampSec(double rampSec) {
```

Comment:
Remember, you won't be able to use those through the interface. Also, you probably don't want to change them while the controller is already active. So these things belong in the consructor.

Reply:
Do try to put default values to not make overwhelming though.

Reply:
Also, you can still keep the function itself for complex set functions like `setCurrentLimits`, just make them private and call them from the ctor only.

Reply:
Why can't the users just call these methods before they call enable()? Too many constructors, in my opinion, because there are not really any default values that will work in every subsystem

Reply:
From what I can tell the default is zero for most of them, that's what will happen if you won't call them.
A lot of ctors is not a reason to make to move to public functions and mutable objects.
I agree that it's annoying that Java doesn't have default function parameters, but you can just use a [builder](https://en.wikipedia.org/wiki/Builder_pattern#Examples).